### PR TITLE
Add auto-scaling support to "add_collection3d"

### LINKED ISSUE #17130: autoscale_view is not working with Line3DCollection
### Bug report

**Bug summary**

I found multiple issues of `autoscale_view` for `Line3DCollection` that 
1. `autoscale_view` takes no effect for plots with `Line3DCollection`.
2. `autoscale_view` even breaks the original bounds and forces it to be a unit cube if it is used with `ax.set_autoscale_on(True)`.

This is reported from the downstream issue https://github.com/sympy/sympy/issues/19106 

**Code for reproduction**

```python
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
from mpl_toolkits.mplot3d.art3d import Line3DCollection

z = np.linspace(-4 * np.pi, 4 * np.pi, 100)
x = np.sin(z)
y = np.cos(z)

segments = []
for i in range(len(z)-1):
    z0, x0, y0 = z[i], x[i], y[i]
    z1, x1, y1 = z[i+1], x[i+1], y[i+1]
    segments.append([(x0, y0, z0), (x1, y1, z1)])

collection = Line3DCollection(segments)
fig = plt.figure()
ax = fig.gca(projection='3d')

ax.add_collection3d(collection)
ax.set_xlim3d(-1, 1)
ax.set_ylim3d(-1, 1)
ax.set_zlim3d(-4*np.pi, 4*np.pi)
ax.set_autoscale_on(True)
ax.autoscale_view()
plt.show()
```

**Actual outcome**

image

**Expected outcome**

image

**Matplotlib version**

  * Operating system:
  * Matplotlib version: 3.2
  * Matplotlib backend (`print(matplotlib.get_backend())`): module://ipykernel.pylab.backend_inline
  * Python version: 3.7.7
  * Jupyter version (if applicable):
  * Other libraries:

### PR DESCRIPTION
**Pull request by lganic**
## PR Summary
closes #17130

Adds support for auto-scaling 3D collection objects in the add_collection3d function. The new implementation is consistent with the behavior of the existing 2D implementation, where auto-scaling is True by default.

Test cases were added for each of the relevant autoscaling functions. 

The autoscaling works with most 2D/3D collections; the only collection I wasn't able to get working was Patch3DCollection, which (as far as I can tell) seems to have a number of ongoing bugs that make them borderline unusable, and untestable. 

Also turns out that a few of the test cases suffered from the bug seen in #17130, so I updated the failing tests accordingly. 
## PR Checklist

**Linked Issue**
- [x] Added "closes #0000" in the PR description to link it to the original issue.

**Documentation and Tests**
- [x] Has pytest style unit tests (and `pytest` passes)
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] New plotting related features are documented with examples.

**Release Notes**
- [x] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [x] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [x] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

**Comment by scottshambaugh** [comment]
I believe this may also close https://github.com/matplotlib/matplotlib/issues/23317, and we should check if it helps https://github.com/matplotlib/matplotlib/issues/14298 as well.

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 1]
```diff
@@ -2188,6 +2191,26 @@ def add_collection3d(self, col, zs=0, zdir='z'):
...
 
+        # FIXME: Implement auto-scaling function for Patch3DCollection
+        # Currently unable to do so due to issues with Patch3DCollection
+        # See issue #14298 for details
+
+        # Bounding box checks
+        # If a 2D collection was passed, it has been converted to 3D
+
+        if autolim:
+            if type(col) is art3d.Line3DCollection:
```
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/axes3d.py:2202]: Is there a reason to not use `isinstance` here?

**Review by QuLogic — CHANGES_REQUESTED** [review]
We would prefer if you could rebase instead of merge. It appears there are some completely unrelated commits in this PR that shouldn't be here.

**Comment by scottshambaugh** [comment]
Hi @lganic, are you still interested in working on this PR?

**Pull request closed without merging**